trie/hasher.go

```diff
@@ -27,26 +27,40 @@ import (
 )
 
 type hasher struct {
-	tmp                  *bytes.Buffer
-	sha                  hash.Hash
 	cachegen, cachelimit uint16
+	threaded             bool
+	mu                   sync.Mutex
 }
 
-// hashers live in a global pool.
-var hasherPool = sync.Pool{
+func newHasher(cachegen, cachelimit uint16) *hasher {
+	h := &hasher{
+		cachegen:   cachegen,
+		cachelimit: cachelimit,
+	}
+	return h
+}
+
+type calculator struct {
+	sha    hash.Hash
+	buffer *bytes.Buffer
+}
```

Comment:
Since this is a utility type used by the hasher, it's cleaner to have it (and its methods) defined at the top of the file. Please bubble them up. Also please add docs for the type to state what it does.